libpdb/client.go

```diff
@@ -50,40 +59,75 @@ func (c *Client) Ping() bool {
 	}
 }
 
-func (c *Client) CreateTopic() (*Topic, error) {
-	password := ""
-	handle, err := NewTopic(password)
-	//@todo
-	return handle, err
-}
+func (c *Client) Publish(handle *Topic, data []byte) error {
+	config := c.config.Load().(ClientConfig)
+	write_args, err := handle.GeneratePublish(config.CommonConfig, data)
+	if err != nil {
+		return err
+	}
 
-func (c *Client) Publish(data []byte) bool {
-	//@todo using EnqueueWrite
-	return true
+	c.msgReqMan.EnqueueWrite(write_args)
+	return nil
 }
 
-func (c *Client) Subscribe(handle *Topic) bool {
-	//@todo using EnqueueRead
-	return true
+func (c *Client) Poll(handle *Subscription) chan []byte {
+	// Check if already subscribed.
+	c.subscriptionMutex.Lock()
+	for x := range c.subscriptions {
+		if &c.subscriptions[x] == handle {
+			c.subscriptionMutex.Unlock()
+			return nil
+		}
+	}
+	c.subscriptions = append(c.subscriptions, *handle)
+	c.subscriptionMutex.Unlock()
+
+	return handle.Updates
 }
 
-func (c *Client) PublishTrace() uint64 {
-	config := c.config.Load().(ClientConfig)
-	req := &common.WriteArgs{}
-	req.ReplyChan = make(chan *common.WriteReply)
-	c.msgReqMan.generateRandomWrite(config, req)
-	c.msgReqMan.EnqueueWrite(req)
-	reply := <-req.ReplyChan
-	return reply.GlobalSeqNo
+func (c *Client) Done(handle *Subscription) bool {
+	c.subscriptionMutex.Lock()
+	for i := 0; i < len(c.subscriptions); i++ {
+		if &c.subscriptions[i] == handle {
+			c.subscriptions[i] = c.subscriptions[len(c.subscriptions)-1]
+			c.subscriptions = c.subscriptions[:len(c.subscriptions)-1]
+			c.subscriptionMutex.Unlock()
+			return true
+		}
+	}
+	c.subscriptionMutex.Unlock()
+	return false
 }
 
-func (c *Client) PollTrace() common.Range {
+// Implement RequestGenerator interface for the request manager
+func (c *Client) NextRequest() (*common.ReadArgs, RequestResponder) {
 	config := c.config.Load().(ClientConfig)
-	req := &common.ReadRequest{}
-	req.Args = &common.ReadArgs{}
-	req.ReplyChan = make(chan *common.ReadReply)
-	c.msgReqMan.generateRandomRead(config, req.Args)
-	c.msgReqMan.EnqueueRead(req)
-	reply := <-req.ReplyChan
-	return reply.GlobalSeqNo
+
+	c.subscriptionMutex.Lock()
+	if c.pendingRequest != nil {
```

Comment:
Instead of using a pendingRequest, what about a queue of requests? Once we integrate interest vectors, we'll need to store more state than the next request anyway